lib/pos_mempool_persister.go

```diff
@@ -0,0 +1,205 @@
+	startGroup sync.WaitGroup
+
+	frequencyMilliseconds int
+	status                MempoolPersisterStatus
+	dbCtx                 *storage.DatabaseContext
+	eventQueue            chan *MempoolEvent
+	updateBatch           []*MempoolEvent
+}
+
+func NewMempoolPersister(dbCtx *storage.DatabaseContext, frequencyMilliseconds int) *MempoolPersister {
+	return &MempoolPersister{
+		frequencyMilliseconds: frequencyMilliseconds,
+		status:                MempoolPersisterStatusNotRunning,
+		dbCtx:                 dbCtx,
+		eventQueue:            make(chan *MempoolEvent, eventQueueSize),
+	}
+}
+
+func (mp *MempoolPersister) Start() {
+	if mp.status == MempoolPersisterStatusRunning {
+		return
+	}
+
+	mp.reset()
+	mp.stopGroup.Add(1)
+	mp.startGroup.Add(1)
+	go mp.run()
+	mp.startGroup.Wait()
+	mp.status = MempoolPersisterStatusRunning
+}
+
+func (mp *MempoolPersister) run() {
+	mp.startGroup.Done()
+	for {
+		select {
+		case event := <-mp.eventQueue:
+			switch event.Type {
+			case MempoolEventAdd, MempoolEventRemove:
+				mp.Lock()
+				mp.updateBatch = append(mp.updateBatch, event)
+				mp.Unlock()
+
+			case MempoolEventExit:
+				mp.status = MempoolPersisterStatusNotRunning
+				mp.stopGroup.Done()
+				return
+			}
+			continue
+
+		case <-time.After(time.Duration(mp.frequencyMilliseconds) * time.Millisecond):
+			if err := mp.persistBatch(); err != nil {
+				glog.Errorf("MempoolPersister: Error persisting batch: %v", err)
+			}
+			continue
+		}
+	}
+}
+
+func (mp *MempoolPersister) Stop() error {
+	if mp.status == MempoolPersisterStatusNotRunning {
+		return nil
+	}
+	mp.eventQueue <- &MempoolEvent{Type: MempoolEventExit}
+	mp.stopGroup.Wait()
+	if err := mp.persistBatch(); err != nil {
+		return errors.Wrapf(err, "MempoolPersister: Error persisting batch")
+	}
+	mp.reset()
+	return nil
+}
+
+func (mp *MempoolPersister) persistBatch() error {
+	if mp.status == MempoolPersisterStatusNotRunning {
+		return nil
+	}
+
+	mp.Lock()
+	defer mp.Unlock()
+
+	if len(mp.updateBatch) == 0 {
+		return nil
+	}
+
+	localContext := mp.dbCtx.GetContext([]byte(DbMempoolContextId))
+	err := mp.dbCtx.Update(localContext, func(txn storage.Transaction, ctx storage.Context) error {
+		for _, event := range mp.updateBatch {
+			if event.Txn == nil || event.Txn.Hash == nil {
+				continue
+			}
+			key := event.Txn.Hash.ToBytes()
+			value, err := event.Txn.ToBytes()
+			if err != nil {
+				continue
+			}
+
+			switch event.Type {
+			case MempoolEventAdd:
+				if err := txn.Set(key, value, ctx); err != nil {
+					glog.Errorf("MempoolPersister: Error setting key: %v", err)
+				}
+			case MempoolEventRemove:
+				if err := txn.Delete(key, ctx); err != nil {
+					glog.Errorf("MempoolPersister: Error deleting key: %v", err)
+				}
+			}
+		}
+		return nil
+	})
+	mp.updateBatch = nil
+
+	if err != nil {
+		return errors.Wrapf(err, "MempoolPersister: Error persisting batch")
+	}
+	return nil
+}
+
+func (mp *MempoolPersister) RetrieveTransactions() ([]*MempoolTx, error) {
```

Comment:
This only returns the persisted transactions (not the ones in the updateBatch). So maybe rename to that to make the usage clearer.

Also, is there ever a scenario where you would want this to return all txns (both from the DB, and from the current un-persisted batch)? Doesn't seem like it since we would really only run this function on node boot-up
```suggestion
func (mp *MempoolPersister) GetPersistedTransactions() ([]*MempoolTx, error) {
```

Reply:
Feels like in such a case mempool will be available and we could use it to retrieve the full view of transactions.